ClusterObserver/Observers/ClusterObserver.cs

```diff
@@ -0,0 +1,131 @@
+﻿// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+using System;
+using System.Fabric;
+using System.Fabric.Health;
+using System.Threading;
+using System.Threading.Tasks;
+using FabricClusterObserver.Utilities.Telemetry;
+using FabricClusterObserver.Utilities;
+
+namespace FabricClusterObserver
+{
+    class ClusterObserver : ObserverBase
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ClusterObserver"/> class.
+        /// This observer runs on one node and as an independent service since FabricObserver 
+        /// is a -1 singleton partition service (runs on every node). ClusterObserver and FabricObserver
+        /// can run in the same cluster as they are independent processes...
+        /// </summary>
+        public ClusterObserver()
+            : base(ObserverConstants.ClusterObserverName)
+        {
+        }
+
+        public override async Task ObserveAsync(CancellationToken token)
+        {
+            // If set, this observer will only run during the supplied interval.
+            // See Settings.xml
+            if (this.RunInterval > TimeSpan.MinValue
+                && DateTime.Now.Subtract(this.LastRunDateTime) < this.RunInterval)
+            {
+                return;
+            }
+
+            if (token.IsCancellationRequested)
+            {
+                this.Token.ThrowIfCancellationRequested();
+
+                return;
+            }
+
+            await ReportAsync(token).ConfigureAwait(true);
+            this.LastRunDateTime = DateTime.Now;
+        }
+
+        public override async Task ReportAsync(CancellationToken token)
+        {
+            await this.ProbeClusterHealthAsync(token).ConfigureAwait(true);
+        }
+
+        private async Task ProbeClusterHealthAsync(CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+
+            _ = bool.TryParse(
+                    this.GetSettingParameterValue(
+                    ObserverConstants.ClusterObserverConfigurationSectionName,
+                    ObserverConstants.EmitHealthWarningEvaluationConfigurationSetting), out bool emitWarningDetails);
+            
+            try
+            {
+                var clusterHealth = await this.FabricClientInstance.HealthManager.GetClusterHealthAsync(
+                                                this.AsyncClusterOperationTimeoutSeconds,
+                                                token).ConfigureAwait(true);
+
+                if (clusterHealth.AggregatedHealthState == HealthState.Ok
```

Comment:
What if we want to report a change from a Warning or Error state which has moved to an Ok state in this evaluation? Wouldn't we still want to report this case to Telemetry?

Reply:
Added configuration option for this